src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs

```diff
@@ -910,5 +948,99 @@ private ICollection<ProjectWithInnerNodes> LoadProjects(IEnumerable<ProjectGraph
+        /// <param name="log">A <see cref="ILogger"/> instance.</param>
+        /// <returns>True if the project has CentralPackageVersionManagement enabled.</returns>
+        internal static bool IsCentralVersionsManagementEnabled(IMSBuildProject project, Common.ILogger log)
+        {
+            var projectStyleResult = BuildTasksUtility.GetProjectRestoreStyle(
+                restoreProjectStyle: project.GetProperty("RestoreProjectStyle"),
+                hasPackageReferenceItems: project.GetItems("PackageReference")?.Any()??false,
+                projectJsonPath: project.GetProperty("_CurrentProjectJsonPath"),
+                projectDirectory: project.Directory,
+                projectName: project.GetProperty("MSBuildProjectName"),
+                log: log);
+
+            return StringComparer.OrdinalIgnoreCase.Equals(project.GetProperty("_CentralPackageVersionsEnabled"), bool.TrueString) &&
+                projectStyleResult.ProjectStyle == ProjectStyle.PackageReference;
+        }
+
+        /// <summary>
+        /// It merges the Versions defined by <paramref name="centralPackageVersions"/> into the metadata of the <paramref name="packageReferences"/>. 
+        /// </summary>
+        /// <param name="packageReferences">The PackageReferences list.</param>
+        /// <param name="centralPackageVersions">The CentralPackageVersion list.</param>
+        internal static void MergeCentralPackageVersionsInPackageReferences(IList<IMSBuildProjectItem> packageReferences,
+            IList<IMSBuildProjectItem> centralPackageVersions)
+        {
+            if (packageReferences == null)
+            {
+                throw new ArgumentNullException(nameof(packageReferences));
+            }
+
+            if (centralPackageVersions == null)
+            {
+                throw new ArgumentNullException(nameof(centralPackageVersions));
+            }
+
+            // Do not touch the IsImplicitlyDefined package references. 
+            var explicitlyReferencedPackages = packageReferences
+                   .Where(item => !IsPropertyTrue(item, "IsImplicitlyDefined"))
+                   .ToList();
+
+            var implicitlyReferencedPackages = packageReferences
+                   .Where(item => IsPropertyTrue(item, "IsImplicitlyDefined"))
+                   .ToList();
+
+            var prWithVersions = string.Join(";", explicitlyReferencedPackages
+                .Where(pr => pr.Properties.Contains("Version") || pr.Properties.Contains("VersionRange"))
+                .Select(pr => pr.Identity));
+            if(!string.IsNullOrEmpty(prWithVersions))
+            {
+                throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, Strings.Error_CentralPackageVersions_VersionsNotAllowed, prWithVersions));
+            }
+
+            var centralVersionAsDictionary = centralPackageVersions
+                .ToDictionary(g => g.Identity.ToLowerInvariant(), g => g);
```

Comment:
You could consider [creating a dictionary with a specific comparer](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.todictionary?view=netframework-4.8#System_Linq_Enumerable_ToDictionary__3_System_Collections_Generic_IEnumerable___0__System_Func___0___1__System_Func___0___2__System_Collections_Generic_IEqualityComparer___1__), so that you can avoid calling `ToLowerInvariant` everywhere. I doubt it would have a performance impact, but it couldn't hurt.

Reply:
Added the suggestion.